src/ui/components/TestSuite/views/TestRecording/TestRecordingEvents/UserActionEventRow.tsx

```diff
@@ -140,7 +140,8 @@ export default memo(function UserActionEventRow({
         }
 
         if (event === userActionEvent) {
-          return eventNumber;
+          // Prepend eventNumber with a zero if it's under 10
+          return eventNumber < 10 ? `0${eventNumber}` : eventNumber;
```

Comment:
![Screenshot 2023-12-07 at 8 42 52 AM](https://github.com/replayio/devtools/assets/29597/7295011f-b2cb-4663-9091-a30508573a77)

More info at https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_counter_styles/Using_CSS_counters

```diff
diff --git a/src/ui/components/TestSuite/views/TestRecording/TestRecordingEvents/UserActionEventRow.module.css b/src/ui/components/TestSuite/views/TestRecording/TestRecordingEvents/UserActionEventRow.module.css
index 964585729..6f74f0b0e 100644
--- a/src/ui/components/TestSuite/views/TestRecording/TestRecordingEvents/UserActionEventRow.module.css
+++ b/src/ui/components/TestSuite/views/TestRecording/TestRecordingEvents/UserActionEventRow.module.css
@@ -27,6 +27,10 @@
   color: var(--color-dim);
   margin-right: 1ch;
 }
+.Number::before {
+  counter-increment: ruler;
+  content: counter(ruler, decimal-leading-zero) " ";
+}
 
 .Name[data-name="assert"] {
   padding: 0 0.25rem;
diff --git a/src/ui/components/TestSuite/views/TestRecording/TestRecordingEvents/UserActionEventRow.tsx b/src/ui/components/TestSuite/views/TestRecording/TestRecordingEvents/UserActionEventRow.tsx
index d20b38753..e6fb1f66b 100644
--- a/src/ui/components/TestSuite/views/TestRecording/TestRecordingEvents/UserActionEventRow.tsx
+++ b/src/ui/components/TestSuite/views/TestRecording/TestRecordingEvents/UserActionEventRow.tsx
@@ -12,7 +12,6 @@ import {
   RecordingTestMetadataV3,
   TestSectionName,
   UserActionEvent,
-  isUserActionTestEvent,
 } from "shared/test-suites/RecordingTestMetadata";
 import { jumpToKnownEventListenerHit } from "ui/actions/eventListeners/jumpToCode";
 import { seek } from "ui/actions/timeline";
@@ -20,9 +19,7 @@ import { JumpToCodeButton, JumpToCodeStatus } from "ui/components/shared/JumpToC
 import { useJumpToSource } from "ui/components/TestSuite/hooks/useJumpToSource";
 import {
   TestEventDetailsEntry,
-  TestEventDomNodeDetails,
   testEventDetailsResultsCache,
-  testEventDomNodeCache,
 } from "ui/components/TestSuite/suspense/TestEventDetailsCache";
 import { TestSuiteContext } from "ui/components/TestSuite/views/TestSuiteContext";
 import { getViewMode } from "ui/reducers/layout";
@@ -124,30 +121,6 @@ export default memo(function UserActionEventRow({
     jumpToCodeStatus = !!jumpToCodeAnnotation ? "found" : "no_hits";
   }
 
-  const eventNumber = useMemo(() => {
-    if (parentId !== null) {
-      return null;
-    }
-
-    let eventNumber = 0;
-
-    const events = testRecording.events[testSectionName];
-    for (let index = 0; index < events.length; index++) {
-      const event = events[index];
-      if (isUserActionTestEvent(event)) {
-        if (event.data.parentId === null) {
-          eventNumber++;
-        }
-
-        if (event === userActionEvent) {
-          return eventNumber;
-        }
-      }
-    }
-
-    return null;
-  }, [parentId, testRecording, testSectionName, userActionEvent]);
-
   return (
     <div
       className={styles.Row}
@@ -158,7 +131,7 @@ export default memo(function UserActionEventRow({
       onMouseLeave={() => setIsHovered(false)}
     >
       <div className={styles.Text}>
-        {eventNumber != null ? <span className={styles.Number}>{eventNumber}</span> : null}
+        <span className={styles.Number} />
         <span
           className={`${styles.Name} ${styles.Name}`}
           data-name={command.name}
diff --git a/src/ui/components/TestSuite/views/TestRecording/TestSection.module.css b/src/ui/components/TestSuite/views/TestRecording/TestSection.module.css
index e82107d94..4d56942ac 100644
--- a/src/ui/components/TestSuite/views/TestRecording/TestSection.module.css
+++ b/src/ui/components/TestSuite/views/TestRecording/TestSection.module.css
@@ -7,3 +7,7 @@
 
   padding-left: 0.75rem;
 }
+
+.Rows {
+  counter-reset: ruler;
+}
diff --git a/src/ui/components/TestSuite/views/TestRecording/TestSection.tsx b/src/ui/components/TestSuite/views/TestRecording/TestSection.tsx
index 21a347163..f20ae366f 100644
--- a/src/ui/components/TestSuite/views/TestRecording/TestSection.tsx
+++ b/src/ui/components/TestSuite/views/TestRecording/TestSection.tsx
@@ -33,14 +33,16 @@ export default function TestSection({
       <div className={styles.Title} data-test-name="TestSection">
         {title}
       </div>
-      {testEvents.map((testEvent, index) => (
-        <TestSectionRow
-          key={index}
-          testEvent={testEvent}
-          testRunnerName={groupedTestCases.environment.testRunner?.name ?? null}
-          testSectionName={testSectionName}
-        />
-      ))}
+      <div className={styles.Rows}>
+        {testEvents.map((testEvent, index) => (
+          <TestSectionRow
+            key={index}
+            testEvent={testEvent}
+            testRunnerName={groupedTestCases.environment.testRunner?.name ?? null}
+            testSectionName={testSectionName}
+          />
+        ))}
+      </div>
     </>
   );
 }
```

Reply:
Never heard of ruler, thanks! Will look into to it.

The zero fill UX debate is an oldie. Even a right-align doesn't work because what if it gets to 1000 steps or 10000? You can't push everything over just in case the steps reach X. There will always be a point where things will move over, and 99->100 is a great place for that. 9->10 less so.

The main question is which looks aesthetically correct, "00" or " 0". This is subjective but here's what I like about 00:

1. It creates only a single layout adjustment, from 99->100. Right-aligning makes a little change from 9->10 and then again at 99->100.

2. For logs/readouts, monospacing with a consistent width looks better. (Subjective opinion) I wouldn't argue for "00" in an academic paper or consumer product marketing site, but for a giant wall of steps for a devtools product that's already using monospaced? Perfect place for zero fill. Looks better and is aligned to the general log aesthetic.